fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,566 @@
+    fn eq(&self, other: &Self) -> bool {
+        self.sig == other.sig
+    }
+}
+
+impl Eq for Secp256r1Signature {}
+
+impl Display for Secp256r1Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{}", Base64::encode(self.as_ref()))
+    }
+}
+
+impl Default for Secp256r1Signature {
+    fn default() -> Self {
+        // Return the signature (1,1)
+        Secp256r1Signature {
+            sig: ExternalSignature::from_scalars(Scalar::ONE.to_bytes(), Scalar::ONE.to_bytes())
+                .unwrap(),
+            bytes: OnceCell::new(),
+            recovery_id: 0u8,
+        }
+    }
+}
+
+// There is a strong requirement for this specific impl. in Fab benchmarks
+/// Secp256r1 public/private key pair.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(tag = "type")] // necessary so as not to deser under a != type
+pub struct Secp256r1KeyPair {
+    pub name: Secp256r1PublicKey,
+    pub secret: Secp256r1PrivateKey,
+}
+
+impl EncodeDecodeBase64 for Secp256r1KeyPair {
+    fn encode_base64(&self) -> String {
+        let mut bytes: Vec<u8> = Vec::new();
+        bytes.extend_from_slice(self.secret.as_ref());
+        bytes.extend_from_slice(self.name.as_ref());
+        Base64::encode(&bytes[..])
+    }
+
+    fn decode_base64(value: &str) -> Result<Self, eyre::Report> {
+        keypair_decode_base64(value)
+    }
+}
+
+impl KeyPair for Secp256r1KeyPair {
+    type PubKey = Secp256r1PublicKey;
+    type PrivKey = Secp256r1PrivateKey;
+    type Sig = Secp256r1Signature;
+
+    fn public(&'_ self) -> &'_ Self::PubKey {
+        &self.name
+    }
+
+    fn private(self) -> Self::PrivKey {
+        Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap()
+    }
+
+    #[cfg(feature = "copy_key")]
+    fn copy(&self) -> Self {
+        Secp256r1KeyPair {
+            name: self.name.clone(),
+            secret: Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap(),
+        }
+    }
+
+    fn generate<R: AllowedRng>(rng: &mut R) -> Self {
+        let privkey = ExternalSecretKey::random(rng);
+        let pubkey = ExternalPublicKey::from(&privkey);
+
+        Secp256r1KeyPair {
+            name: Secp256r1PublicKey {
+                pubkey,
+                bytes: OnceCell::new(),
+            },
+            secret: Secp256r1PrivateKey {
+                privkey,
+                bytes: OnceCell::new(),
+            },
+        }
+    }
+}
+
+impl FromStr for Secp256r1KeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl Signer<Secp256r1Signature> for Secp256r1KeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Secp256r1Signature, signature::Error> {
+        // Code copied from Sign.rs in k256@0.11.6
+
+        // Hash message
+        let z = FieldBytes::from(Sha256::digest(msg).digest);
```

Comment:
Interesting that internally the message is hashed with sha256.

Reply:
Yes, I noticed that too. However, now that we have reproduced all the code, we could easily make the hash function generic.